OptimizelySDK.Tests/EventTests/BatchEventProcessorTest.cs

```diff
@@ -0,0 +1,272 @@
+        private TestEventDispatcher TestEventDispatcher;
+
+        private NotificationCenter NotificationCenter = new NotificationCenter();
+        private Mock<TestNotificationCallbacks> NotificationCallbackMock;
+
+        [SetUp]
+        public void Setup()
+        {
+            LoggerMock = new Mock<ILogger>();
+            LoggerMock.Setup(l => l.Log(It.IsAny<LogLevel>(), It.IsAny<string>()));
+
+            Config = DatafileProjectConfig.Create(TestData.Datafile, LoggerMock.Object, new ErrorHandler.NoOpErrorHandler());
+            
+            eventQueue = new BlockingCollection<object>(100);
+            EventDispatcherMock = new Mock<IEventDispatcher>();
+
+            NotificationCallbackMock = new Mock<TestNotificationCallbacks>();
+            NotificationCallbackMock.Setup(nc => nc.TestLogEventCallback(It.IsAny<LogEvent>()));
+
+            NotificationCenter.AddNotification(NotificationCenter.NotificationType.LogEvent,
+                NotificationCallbackMock.Object.TestLogEventCallback);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            EventProcessor.Stop();
+        }
+        
+        [Test]
+        public void TestDrainOnClose()
+        {
+            var eventDispatcher = new TestEventDispatcher();
+            SetEventProcessor(eventDispatcher);
+
+            UserEvent userEvent = BuildConversionEvent(EventName);
+            EventProcessor.Process(userEvent);
+            eventDispatcher.ExpectConversion(EventName, TestUserId);
+
+            Thread.Sleep(500);
+
+            Assert.True(eventDispatcher.CompareEvents());
+            Assert.AreEqual(0, EventProcessor.EventQueue.Count);
+        }
+
+        [Test]
+        public void TestFlushOnMaxTimeout()
+        {
+            var countdownEvent = new CountdownEvent(1);
+            var eventDispatcher = new TestEventDispatcher(countdownEvent);
+            SetEventProcessor(eventDispatcher);
+
+            UserEvent userEvent = BuildConversionEvent(EventName);
+            EventProcessor.Process(userEvent);
+            eventDispatcher.ExpectConversion(EventName, TestUserId);
+
+            Thread.Sleep(500);
+
+            Assert.True(eventDispatcher.CompareEvents());
+            Assert.True(countdownEvent.Wait(TimeSpan.FromMilliseconds(MAX_DURATION_MS * 3)), "Exceeded timeout waiting for notification.");
+            Assert.AreEqual(0, EventProcessor.EventQueue.Count);
+        }
+
+        [Test]
+        public void TestFlushMaxBatchSize()
+        {
+            var countdownEvent = new CountdownEvent(1);
+            var eventDispatcher = new TestEventDispatcher(countdownEvent);
+            SetEventProcessor(eventDispatcher);
+
+            for (int i = 0; i < MAX_BATCH_SIZE; i++)
+            {
+                UserEvent userEvent = BuildConversionEvent(EventName);
+                EventProcessor.Process(userEvent);
+                eventDispatcher.ExpectConversion(EventName, TestUserId);
```

Comment:
Hmm, so with the batch processor, we expect the userEvents to be grouped up to the `MAX_BATCH_SIZE` and then only one `LogEvent` should make it to the eventDispatcher on flush

Reply:
Yes, I also added logger verification in the unit test to validate that DispatchEvent gets called only once.